include/aie/Dialect/AIEVec/IR/AIEVecTypeConstraints.td

```diff
@@ -144,4 +144,64 @@ class IsValidAIE2MatMulShapeAndType<string lhs, string rhs, string acc> :
+            # "accumulate",
+            "::mlir::VectorType">;
+
+def AIE2MulElemRHS :
+  AnyTypeOf<[VectorOfShapeAndType<[32], I8>,
+             VectorOfShapeAndType<[32], I16>,
+             VectorOfShapeAndType<[16], I32>,
+             VectorOfShapeAndType<[16], BF16>,
+             VectorOfShapeAndType<[16], F32>],
+            "a vector compatible with a rhs operand of element-wise multiply and "
+            # "accumulate",
+            "::mlir::VectorType">;
+
+def AIE2MulElemACC :
+  AnyTypeOf<[VectorOfShapeAndType<[32], I8>,
+             VectorOfShapeAndType<[32], I16>,
+             VectorOfShapeAndType<[16], I32>,
+             VectorOfShapeAndType<[32], I32>,
+             VectorOfShapeAndType<[16], BF16>,
+             VectorOfShapeAndType<[16], F32>],
+            "a vector compatible with an accumulator of element-wise multiply and "
+            # "accumulate",
+            "::mlir::VectorType">;
+
+class IsValidAIE2MulElemShapeAndType<string lhs, string rhs, string acc> :
+  PredOpTrait<lhs # " x " # rhs # " = " # acc # " is a valid AIE2 " #
+              "element-wise multiply and accumulate op",
+              Or<[VectorTypesMatch<lhs, VectorOfShapeAndType<[32], I8>,
+                                   rhs, VectorOfShapeAndType<[32], I8>,
+                                   acc, VectorOfShapeAndType<[32], I8>>,
+                  VectorTypesMatch<lhs, VectorOfShapeAndType<[32], I8>,
+                                   rhs, VectorOfShapeAndType<[32], I8>,
+                                   acc, VectorOfShapeAndType<[32], I32>>,
+                  VectorTypesMatch<lhs, VectorOfShapeAndType<[32], I16>,
+                                   rhs, VectorOfShapeAndType<[32], I16>,
+                                   acc, VectorOfShapeAndType<[32], I16>>,
+                  VectorTypesMatch<lhs, VectorOfShapeAndType<[32], I16>,
+                                   rhs, VectorOfShapeAndType<[32], I16>,
+                                   acc, VectorOfShapeAndType<[32], I32>>,
+
+                  VectorTypesMatch<lhs, VectorOfShapeAndType<[16], I32>,
+                                   rhs, VectorOfShapeAndType<[16], I32>,
+                                   acc, VectorOfShapeAndType<[16], I32>>,
+                  VectorTypesMatch<lhs, VectorOfShapeAndType<[16], BF16>,
+                                   rhs, VectorOfShapeAndType<[16], BF16>,
+                                   acc, VectorOfShapeAndType<[16], BF16>>,
+                  VectorTypesMatch<lhs, VectorOfShapeAndType<[16], BF16>,
+                                   rhs, VectorOfShapeAndType<[16], BF16>,
+                                   acc, VectorOfShapeAndType<[16], F32>>,
+                  VectorTypesMatch<lhs, VectorOfShapeAndType<[16], F32>,
+                                   rhs, VectorOfShapeAndType<[16], F32>,
+                                   acc, VectorOfShapeAndType<[16], F32>>]>>;
```

Comment:
This constraint alone already makes all the checks necessary for the supported types, but it's unnecessarily complex for such a regular operation.